include/ocpp/common/safe_queue.hpp

```diff
@@ -0,0 +1,118 @@
+
+public:
+    /// \return True if the queue is empty
+    inline bool empty() const {
+        std::lock_guard lock(mutex);
+        return queue.empty();
+    }
+
+    inline safe_queue_reference front() const {
+        std::lock_guard lock(mutex);
+        return queue.front();
+    }
+
+    inline safe_queue_const_reference front() {
+        std::lock_guard lock(mutex);
+        return queue.front();
+    }
+
+    /// \return retrieves and removes the first element in the queue. Undefined behavior if the queue is empty
+    inline T pop() {
+        std::lock_guard lock(mutex);
+
+        T front = std::move(queue.front());
+        queue.pop();
+
+        return front;
+    }
+
+    /// \brief Queues an element and notifies any threads waiting on the internal conditional variable
+    inline void push(T&& value) {
+        {
+            std::lock_guard<std::mutex> lock(mutex);
+            queue.push(value);
+        }
+
+        notify_waiting_thread();
+    }
+
+    /// \brief Queues an element and notifies any threads waiting on the internal conditional variable
+    inline void push(const T& value) {
+        {
+            std::lock_guard<std::mutex> lock(mutex);
+            queue.push(value);
+        }
+
+        notify_waiting_thread();
+    }
+
+    /// \brief Clears the queue
+    inline void clear() {
+        std::lock_guard<std::mutex> lock(mutex);
+
+        std::queue<T> empty;
+        empty.swap(queue);
+    }
+
+    /// \brief Waits seconds for the queue to receive an element
+    /// \param seconds Count of seconds to wait, pass in a value <= 0 to wait indefinitely
+    inline void wait_on_queue(int seconds = -1) {
+        std::unique_lock<std::mutex> lock(mutex);
+
+        if (seconds > 0) {
+            cv.wait_for(lock, std::chrono::seconds(seconds), [&]() { return (false == queue.empty()); });
+        } else {
+            cv.wait(lock, [&]() { return (false == queue.empty()); });
+        }
+    }
+
+    /// \brief Same as 'wait_on_queue' but receives an additional predicate to wait upon
+    template <class Predicate> inline void wait_on_queue(Predicate pred, int seconds = -1) {
+        std::unique_lock<std::mutex> lock(mutex);
+
+        if (seconds > 0) {
+            cv.wait_for(lock, std::chrono::seconds(seconds), [&]() { return (false == queue.empty()) or pred(); });
+        } else {
+            cv.wait(lock, [&]() { return (false == queue.empty()) or pred(); });
+        }
+    }
+
+    /// \brief Waits on the queue for a custom event
+    template <class Predicate> inline void wait_on_custom(Predicate pred, int seconds = -1) {
+        std::unique_lock<std::mutex> lock(mutex);
+
+        if (seconds > 0) {
+            cv.wait_for(lock, std::chrono::seconds(seconds), [&]() { return pred(); });
+        } else {
+            cv.wait(lock, [&]() { return pred(); });
+        }
+    }
+
+    /// \brief Notifies a single waiting thread to wake up
+    inline void notify_waiting_thread() {
```

Comment:
Any suggestions? Expose the `cv`?

Reply:
I am not sure, it's probably not an easy thing to do right so it's probably best to continue with this for now.

To elaborate on my concern. There are 2 things we are doing; 
1. We have a queue that we need to protect for multi threaded access, that is done in this class. 
2. We have a thread running that is waiting for input from the queue or input from other places. This is now done half in this class (the `cv`) and half in the thread part of the websocket. So either we fully move the thread into this class and handle it only in here, or we move the `cv` out of this one. This has more implications I know, that's why I think its best to leave it as is for now.

Reply:
I do understand the concerns. However, I've thought at something. The problem is that not always the queue knows when to wake up the waiting thread. When we push a message? When we pop a message? Both? That's why the 'notify_waiting_thread' could be removed and replaced with a C++ template policy enum that specifies the used case. 

For example:

`
enum ESafeQueueWakeupPolicy {
WAKE_UP_NEVER,
WAKE_UP_ON_POP,
WAKE_UP_ON_PUSH,
WAKE_UP_ALWAYS
};
`

Of course that adds the complexity of having to know what the queue will do.

For moving the thread in the queue, I've also thought at that but I wouldn't do it in the `SafeQueue` class, I'd create a `QueueWorkerThread` that contains both the `std::thread` and an internal `SafeQueue` and that receives as a ctor a lambda worked parameter or something along those lines.

However at the moment it seems the queue wakeup policy makes the most sense.

Reply:
I am not really sure how that would look yet, might be a good option. At the same time, I understood this needs to be merged soon for some kind of release so might be wise to just accept that this is some technical debt and accept it for now?

Reply:
It would not be that complicated (~30 mins of work), and it would remove the 'notify_waiting_thread()' function better separating functionality as you've said. That will give us a head-start when we will properly fix the other technical debt, the proper management of the thread that is attached to this queue.

Reply:
I've implemented here a queue policy on when to wake up. However I am not sure that it is the best idea. Maybe we should notify the thread always to wake up on an operation?

https://github.com/EVerest/libocpp/pull/896/commits/4ac4485db2d6091c80e28a9b0280d74570a0af14

Reply:
I dont see any drawback with notifying on each queue operation and then we could omit the policy. I would suggest to accept the changes for now.

Of course one could reduce the scope and complexity of the `SafeQueue` class to the basic queue operations, move the cv out of it and pass e.g. a lambda triggered on every operation (or even based on a passed policy), and we can discuss this afterwards.

But in the current design of the libwebsockets impl the logic that uses the queues requires the notifications and this leads to the fact that the SafeQueue always need the notify mechanism one or the other way.  Therefore I'm also fine with the current implementation since the `SafeQueue` is currently only used here

Reply:
At the moment I've removed the policy and will just notify the thread on all operations. The old behavior was that the CV is out of the class, however that added a lot of failure points in the libwebsocket class that always required to call that CV notify, since in there we have 3 queues.

Maybe the lambda triggered on each operation could be the best idea, but I'd say to postpone it until we need the SafeQueue in another class that is not the websocket.